src/mplhep/plot.py

```diff
@@ -550,68 +550,126 @@ def iterable_not_string(arg):
+            ):
+                xticks = _xticks
+                xticklabels = _xticklabels
+                break
+            elif len(_xticklabels) > 0:
+                xticks = _xticks
+                xticklabels = _xticklabels
+
         lw = ax.spines["bottom"].get_linewidth()
         _edges = plottables[0].edges
         _centers = plottables[0].centers
         _marker_size = (
             20
             * ax.get_window_extent().transformed(fig.dpi_scale_trans.inverted()).width
         )
-        if underflow > 0.0:
-            xticks[0] = ""
-            xticks[1] = f"<{flow_bins[2]}"
-            ax.set_xticks(ax.get_xticks())
-            ax.set_xticklabels(xticks)
 
-            ax.plot(
-                [_edges[0], _edges[1]],
-                [0, 0],
-                color="white",
-                zorder=5,
-                ls="--",
-                lw=lw,
-                transform=ax.get_xaxis_transform(),
-                clip_on=False,
-            )
-            ax.scatter(
-                _centers[0],
-                0,
-                _marker_size,
-                marker=align_marker("d", valign="center"),
-                edgecolor="black",
-                zorder=5,
-                clip_on=False,
-                facecolor="white",
-                transform=ax.get_xaxis_transform(),
-            )
-        if overflow > 0.0:
-            xticks[-1] = ""
-            xticks[-2] = f">{flow_bins[-3]}"
-            ax.set_xticks(ax.get_xticks())
-            ax.set_xticklabels(xticks)
-            ax.plot(
-                [_edges[-2], _edges[-1]],
-                [0, 0],
-                color="white",
-                zorder=5,
-                ls="--",
-                lw=lw,
-                transform=ax.get_xaxis_transform(),
-                clip_on=False,
-            )
-            ax.scatter(
-                _centers[-1],
-                0,
-                _marker_size,
-                marker=align_marker("d", valign="center"),
-                edgecolor="black",
-                zorder=5,
-                clip_on=False,
-                facecolor="white",
-                transform=ax.get_xaxis_transform(),
-            )
+        if underflow > 0.0 or underflow_xticklabel in xticklabels:
+            # Replace any existing xticks in underflow region with underflow bin center
+            _mask = xticks > flow_bins[1]
+            xticks = np.insert(xticks[_mask], 0, _centers[0])
+            xticklabels = [underflow_xticklabel] + [
+                xlab for i, xlab in enumerate(xticklabels) if _mask[i]
+            ]
+
+            # Don't draw markers on the top of the top axis
+            top_axis = max(shared_axes, key=lambda a: a.get_position().y0)
+
+            # Draw on all shared axes
+            for _ax in shared_axes:
+                for h in [0, 1]:
+                    _ax.set_xticks(xticks)
+                    _ax.set_xticklabels(xticklabels)
+
```

Comment:
If you move the skip condition up here, wouldn't that prevent drawing the dashed line on top?